File: backend/go/internal/service/customer_service.go
package service

import (
	"fmt"
	"log"

	"github.com/google/uuid"

	"beauty-salon-backend-go/internal/models"
	"beauty-salon-backend-go/internal/repository"
)

// CustomerService handles business logic for customers
type CustomerService struct {
	repo *repository.CustomerRepository
}

// NewCustomerService creates a new customer service
func NewCustomerService(repo *repository.CustomerRepository) *CustomerService {
	return &CustomerService{repo: repo}
}

// CreateCustomer creates a new customer
func (s *CustomerService) CreateCustomer(name, email, phone string) (*models.Customer, error) {
	// Validate input
	if name == "" {
		return nil, models.ErrInvalidName
	}
	if email == "" {
		return nil, models.ErrInvalidEmail
	}

	// Check if customer with email already exists
	existing, err := s.repo.FindByEmail(email)
	if err != nil && err != models.ErrNotFound {
		return nil, fmt.Errorf("failed to check existing customer: %w", err)
	}
	if existing != nil {
		return nil, models.ErrAlreadyExists
	}

	// Create new customer
	customer := models.NewCustomer(name, email, phone)
	
	if err := s.repo.Create(customer); err != nil {
		return nil, fmt.Errorf("failed to create customer: %w", err)
	}

	log.Printf("Created customer: %s (ID: %s)", customer.Name, customer.ID)
	return customer, nil
}

// GetCustomerByID retrieves a customer by ID
func (s *CustomerService) GetCustomerByID(id uuid.UUID) (*models.Customer, error) {
	customer, err := s.repo.FindByID(id)
	if err != nil {
		if err == models.ErrNotFound {
			return nil, models.ErrCustomerNotFound
		}
		return nil, fmt.Errorf("failed to get customer: %w", err)
	}
	return customer, nil
}

// GetCustomerByEmail retrieves a customer by email
func (s *CustomerService) GetCustomerByEmail(email string) (*models.Customer, error) {
	if email == "" {
		return nil, models.ErrInvalidEmail
	}

	customer, err := s.repo.FindByEmail(email)
	if err != nil {
		if err == models.ErrNotFound {
			return nil, models.ErrCustomerNotFound
		}
		return nil, fmt.Errorf("failed to get customer by email: %w", err)
	}
	return customer, nil
}

// GetAllCustomers retrieves all customers
func (s *CustomerService) GetAllCustomers() ([]*models.Customer, error) {
	customers, err := s.repo.FindAll()
	if err != nil {
		return nil, fmt.Errorf("failed to get all customers: %w", err)
	}
	return customers, nil
}

// SearchCustomersByName searches customers by name
func (s *CustomerService) SearchCustomersByName(name string) ([]*models.Customer, error) {
	if name == "" {
		return []*models.Customer{}, nil
	}

	customers, err := s.repo.SearchByName(name)
	if err != nil {
		return nil, fmt.Errorf("failed to search customers: %w", err)
	}
	return customers, nil
}

// UpdateCustomer updates an existing customer
func (s *CustomerService) UpdateCustomer(id uuid.UUID, name, email, phone string) (*models.Customer, error) {
	// Validate input
	if name == "" {
		return nil, models.ErrInvalidName
	}
	if email == "" {
		return nil, models.ErrInvalidEmail
	}

	// Check if customer exists
	customer, err := s.repo.FindByID(id)
	if err != nil {
		if err == models.ErrNotFound {
			return nil, models.ErrCustomerNotFound
		}
		return nil, fmt.Errorf("failed to find customer: %w", err)
	}

	// Check if email is being changed and if new email already exists
	if customer.Email != email {
		existing, err := s.repo.FindByEmail(email)
		if err != nil && err != models.ErrNotFound {
			return nil, fmt.Errorf("failed to check existing customer: %w", err)
		}
		if existing != nil && existing.ID != id {
			return nil, models.ErrAlreadyExists
		}
	}

	// Update customer fields
	customer.Name = name
	customer.Email = email
	customer.Phone = phone
	customer.UpdateTimestamp()

	if err := s.repo.Update(customer); err != nil {
		return nil, fmt.Errorf("failed to update customer: %w", err)
	}

	log.Printf("Updated customer: %s (ID: %s)", customer.Name, customer.ID)
	return customer, nil
}

// DeleteCustomer deletes a customer by ID
func (s *CustomerService) DeleteCustomer(id uuid.UUID) error {
	// Check if customer exists
	exists, err := s.repo.Exists(id)
	if err != nil {
		return fmt.Errorf("failed to check customer existence: %w", err)
	}
	if !exists {
		return models.ErrCustomerNotFound
	}

	if err := s.repo.Delete(id); err != nil {
		return fmt.Errorf("failed to delete customer: %w", err)
	}

	log.Printf("Deleted customer with ID: %s", id)
	return nil
}

// CustomerExists checks if a customer exists by ID
func (s *CustomerService) CustomerExists(id uuid.UUID) (bool, error) {
	exists, err := s.repo.Exists(id)
	if err != nil {
		return false, fmt.Errorf("failed to check customer existence: %w", err)
	}
	return exists, nil
}

// GetCustomerCount gets the total count of customers
func (s *CustomerService) GetCustomerCount() (int, error) {
	count, err := s.repo.Count()
	if err != nil {
		return 0, fmt.Errorf("failed to get customer count: %w", err)
	}
	return count, nil
}

// SearchCustomers searches customers by name, email or phone
func (s *CustomerService) SearchCustomers(query string, limit int) ([]*models.Customer, error) {
	if query == "" {
		return []*models.Customer{}, nil
	}

	customers, err := s.repo.Search(query, limit)
	if err != nil {
		return nil, fmt.Errorf("failed to search customers: %w", err)
	}
	return customers, nil
}

File: backend/python/app/services/customer_service.py
"""
Customer service for Beauty Salon Management System
"""

import logging
from typing import List, Optional
from uuid import UUID
from fastapi import HTTPException, status

from app.models.customer import Customer, CustomerCreate, CustomerUpdate
from app.repositories.customer_repository import CustomerRepository


logger = logging.getLogger(__name__)


class CustomerService:
    """Service for customer business logic"""
    
    def __init__(self, customer_repository: CustomerRepository):
        self.customer_repository = customer_repository
    
    async def create_customer(self, customer_data: CustomerCreate) -> Customer:
        """Create a new customer with business validation"""
        try:
            # Check if email already exists
            if await self.customer_repository.exists_by_email(customer_data.email):
                raise HTTPException(
                    status_code=status.HTTP_409_CONFLICT,
                    detail="Customer with this email already exists"
                )
            
            # Create customer
            customer = await self.customer_repository.create(customer_data)
            logger.info(f"Customer created successfully: {customer.id}")
            return customer
            
        except HTTPException:
            raise
        except Exception as e:
            logger.error(f"Error creating customer: {e}")
            raise HTTPException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                detail="Failed to create customer"
            )
    
    async def get_customer_by_id(self, customer_id: UUID) -> Customer:
        """Get customer by ID"""
        try:
            customer = await self.customer_repository.get_by_id(customer_id)
            if not customer:
                raise HTTPException(
                    status_code=status.HTTP_404_NOT_FOUND,
                    detail="Customer not found"
                )
            return customer
            
        except HTTPException:
            raise
        except Exception as e:
            logger.error(f"Error getting customer by ID {customer_id}: {e}")
            raise HTTPException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                detail="Failed to get customer"
            )
    
    async def get_customer_by_email(self, email: str) -> Customer:
        """Get customer by email"""
        try:
            customer = await self.customer_repository.get_by_email(email)
            if not customer:
                raise HTTPException(
                    status_code=status.HTTP_404_NOT_FOUND,
                    detail="Customer not found"
                )
            return customer
            
        except HTTPException:
            raise
        except Exception as e:
            logger.error(f"Error getting customer by email {email}: {e}")
            raise HTTPException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                detail="Failed to get customer"
            )
    
    async def get_all_customers(self, limit: int = 100) -> List[Customer]:
        """Get all customers"""
        try:
            customers = await self.customer_repository.get_all(limit)
            return customers
            
        except Exception as e:
            logger.error(f"Error getting all customers: {e}")
            raise HTTPException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                detail="Failed to get customers"
            )
    
    async def update_customer(self, customer_id: UUID, customer_data: CustomerUpdate) -> Customer:
        """Update customer with business validation"""
        try:
            # Check if customer exists
            existing_customer = await self.customer_repository.get_by_id(customer_id)
            if not existing_customer:
                raise HTTPException(
                    status_code=status.HTTP_404_NOT_FOUND,
                    detail="Customer not found"
                )
            
            # If email is being updated, check if it's already taken by another customer
            if customer_data.email and customer_data.email != existing_customer.email:
                if await self.customer_repository.exists_by_email(customer_data.email, exclude_id=customer_id):
                    raise HTTPException(
                        status_code=status.HTTP_409_CONFLICT,
                        detail="Email already exists for another customer"
                    )
            
            # Update customer
            updated_customer = await self.customer_repository.update(customer_id, customer_data)
            if not updated_customer:
                raise HTTPException(
                    status_code=status.HTTP_404_NOT_FOUND,
                    detail="Customer not found"
                )
            
            logger.info(f"Customer updated successfully: {customer_id}")
            return updated_customer
            
        except HTTPException:
            raise
        except Exception as e:
            logger.error(f"Error updating customer {customer_id}: {e}")
            raise HTTPException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                detail="Failed to update customer"
            )
    
    async def delete_customer(self, customer_id: UUID) -> bool:
        """Delete customer"""
        try:
            # Check if customer exists
            existing_customer = await self.customer_repository.get_by_id(customer_id)
            if not existing_customer:
                raise HTTPException(
                    status_code=status.HTTP_404_NOT_FOUND,
                    detail="Customer not found"
                )
            
            # TODO: Check if customer has active appointments before deletion
            # This would require appointment repository dependency
            
            # Delete customer
            deleted = await self.customer_repository.delete(customer_id)
            if deleted:
                logger.info(f"Customer deleted successfully: {customer_id}")
            
            return deleted
            
        except HTTPException:
            raise
        except Exception as e:
            logger.error(f"Error deleting customer {customer_id}: {e}")
            raise HTTPException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                detail="Failed to delete customer"
            )
    
    async def search_customers(self, name: str, limit: int = 50) -> List[Customer]:
        """Search customers by name"""
        try:
            if not name or not name.strip():
                raise HTTPException(
                    status_code=status.HTTP_400_BAD_REQUEST,
                    detail="Search term cannot be empty"
                )
            
            customers = await self.customer_repository.search_by_name(name.strip(), limit)
            return customers
            
        except HTTPException:
            raise
        except Exception as e:
            logger.error(f"Error searching customers by name {name}: {e}")
            raise HTTPException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                detail="Failed to search customers"
            )
    
    async def get_customer_count(self) -> int:
        """Get total count of customers"""
        try:
            count = await self.customer_repository.count()
            return count
            
        except Exception as e:
            logger.error(f"Error getting customer count: {e}")
            raise HTTPException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                detail="Failed to get customer count"
            )
    
    async def customer_exists(self, customer_id: UUID) -> bool:
        """Check if customer exists"""
        try:
            customer = await self.customer_repository.get_by_id(customer_id)
            return customer is not None
            
        except Exception as e:
            logger.error(f"Error checking customer existence {customer_id}: {e}")
            return False
